design/ceph-volume-provisioning.md

```diff
@@ -0,0 +1,136 @@
+        dataVolumeGroup: data_vg
+      # bluestore: the DB, WAL, and Data are all on the same LV
+      - volume: my_lv1
+        volumeGroup: my_vg
+      # filestore: data and journal on the same LV
+      - data: my_lv2
+        dataVolumeGroup: my_vg
+      # filestore: data and journal on different LVs
+      - data: data_lv3
+        dataVolumeGroup: data_vg
+        journal: journal_lv3
+        journalVolumeGroup: journal_vg
+      # devices support both filestore and bluestore configurations based on the "config.storeType" setting at the global, node, or device level
+      devices:
+      # OSD on a raw device
+      - name: sdd
+      # OSD on a partition (partition support is new)
+      - name: sdf1
+      # Multiple OSDs on a high performance device
+      - name: nvme01
+        count: 3
+```
+
+The above options for LVM and partitions look very tedious. Questions:
+- Is it useful at this level of complexity?
+- Is there a simpler way users would configure LVM?
+- Do users need all this flexibility? This looks like too many options to maintain.
+
+### Backward compatibility
+
+Rook will need to continue supporting clusters that are running different types of OSDs. All of the v0.8 OSDs must continue running
+after Rook is upgraded to v0.9 and beyond, whether they were filestore or bluestore running on directories or devices.
+
+Since `ceph-volume` only supports devices that have **not** been previously configured by Rook:
+- Rook will continue to provision OSDs directly when a `directory` is specified in the CRD
+  - Support for creating new OSDs on directories will be deprecated. While directories might still be used for test scenarios,
+  it's not a mainline scenario. With the legacy design, directories were commonly used on LVM, but LVM is now directly supported.
+  In v0.9, support for directories will remain, but documentation will encourage users to provision devices.
+- For existing devices configured by Rook, `ceph-volume` will be skipped and the OSDs will be started as previously
+- New devices will be provisioned with `ceph-volume`
+
+### Versioning
+
+Rook relies on very recent developments in `ceph-volume` that are not yet available in luminous or mimic releases.
+For example, rook needs to run the command:
+```
+ceph-volume lvm batch --prepare <devices>
+```
+
+The `batch` command and the flag `--prepare` have been added recently.
+While the latest ceph-volume changes will soon be merged to luminous and mimic, Rook needs to know if it is running an image that contains the required functionality.
+
+To detect if ceph-volume supports the required options, Rook will run the
+command with all the flags that are required. To avoid side effects when testing for the version of ceph-volume, no devices
+are passed to the `batch` command.
+```
+ceph-volume lvm batch --prepare
+```
+
+- If the flags are supported, ceph-volume has an exit code of `0`.
+- If the flags are not supported, ceph-volume has an exit code of `2`.
```

Comment:
This need makes sense, but I think it is going to be hard to return a `0` exit status on an error condition (missing inputs).

The flags and positional arguments are hard-wired into the argument parser library, which takes care of the error handling (e.g. "this flag requires a value to be an int, error otherwise"). I don't think we can easily hijack it to check if the flags passed in are supported to report a `0` exist status (vs. invalid flags). In any case, this might take some effort